compile/constant_value.go

```diff
@@ -76,24 +83,110 @@ type (
 )
 
 // Link for ConstantBool
-func (c ConstantBool) Link(scope Scope) (ConstantValue, error) {
-	return c, nil
+func (c ConstantBool) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	return typeEqualsOrCastError(c, BoolSpec, t)
 }
 
 // Link for ConstantInt.
-func (c ConstantInt) Link(scope Scope) (ConstantValue, error) {
-	// TODO ConstantInt can resolve to a ConstantBool if it's being treated as
-	// a bool.
-	return c, nil
+func (c ConstantInt) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	switch t {
+	case I8Spec, I16Spec, I32Spec, I64Spec:
+		// TODO bounds checks?
+		return c, nil
+	case DoubleSpec:
+		return ConstantDouble(float64(c)).Link(scope, t)
+	case BoolSpec:
+		v := int64(c)
+		switch v {
+		case 0, 1:
+			return ConstantBool(v == 1).Link(scope, t)
+		default:
+			return nil, constantValueCastError{
+				Value:  c,
+				Type:   t,
+				Reason: errors.New("the value must be 0 or 1"),
+			}
+		}
+	default:
+		// fall through
+	}
+
+	// Used for an enum
+	if e, ok := t.(*EnumSpec); ok {
+		for _, item := range e.Items {
+			if item.Value == int32(c) {
+				return EnumItemReference{Enum: e, Item: item}, nil
+			}
+		}
+
+		return nil, constantValueCastError{
+			Value: c,
+			Type:  t,
+			Reason: fmt.Errorf(
+				"%v is not a valid value for enum %q", int32(c), e.ThriftName()),
+		}
+	}
+
+	return nil, constantValueCastError{Value: c, Type: t}
+	// TODO: AST for constants will need to track positions for us to
+	// include them in the error messages.
 }
 
 // Link for ConstantString.
-func (c ConstantString) Link(scope Scope) (ConstantValue, error) {
-	return c, nil
+func (c ConstantString) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	return typeEqualsOrCastError(c, StringSpec, t)
+	// TODO(abg): Are binary literals a thing?
 }
 
 // Link for ConstantDouble.
-func (c ConstantDouble) Link(scope Scope) (ConstantValue, error) {
+func (c ConstantDouble) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	return typeEqualsOrCastError(c, DoubleSpec, t)
+}
+
+// ConstantStruct represents a struct literal from the Thrift file.
+type ConstantStruct struct {
+	Fields map[string]ConstantValue
+}
+
+// Link for ConstantStruct
+func (c *ConstantStruct) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	s, ok := t.(*StructSpec)
+	if !ok {
+		return nil, constantValueCastError{Value: c, Type: t}
+	}
+
+	for _, field := range s.Fields {
+		f, ok := c.Fields[field.Name]
+		if !ok {
+			if field.Default == nil {
+				if field.Required {
+					return nil, constantValueCastError{
+						Value:  c,
+						Type:   t,
+						Reason: fmt.Errorf("%q is a required field", field.Name),
+					}
+				}
+				continue
+			}
+			f = field.Default
+			c.Fields[field.Name] = f
```

Comment:
Do you need to do this? It loses information about whether the user specified an override which happened to be the default, and the user of the library would never be able to check that right?